src/lib/pure_pursuit/PurePursuit.cpp

```diff
@@ -0,0 +1,112 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "PurePursuit.hpp"
+#include <mathlib/mathlib.h>
+
+float PurePursuit::calcDesiredHeading(const Vector2f &curr_wp_ned, const Vector2f &prev_wp_ned,
+				      const Vector2f &curr_pos_ned,
+				      const float &lookahead_distance)
+{
+	// Check minimal necessary input validity
+	if (!curr_wp_ned.isAllFinite() || !curr_pos_ned.isAllFinite()) {
+		return 0.f; // No fallback possible
+	}
+
+	// Setup fallback to bearing to the current waypoint
+	const Vector2f curr_pos_to_curr_wp = curr_wp_ned - curr_pos_ned;
+	const float bearing_to_curr_wp = atan2f(curr_pos_to_curr_wp(1), curr_pos_to_curr_wp(0));
+
+	// Check full input validity
+	if (lookahead_distance < FLT_EPSILON || !PX4_ISFINITE(lookahead_distance) || !prev_wp_ned.isAllFinite()) {
+		return bearing_to_curr_wp; // Fallback
+	}
+
+	// Catch waypoint overlap
+	if (fabsf(curr_wp_ned(0) - prev_wp_ned(0)) < FLT_EPSILON
+	    && fabsf(curr_wp_ned(1) - prev_wp_ned(1)) < FLT_EPSILON) {
+		return bearing_to_curr_wp; // Fallback
+	}
+
+	// Catch infinite line segment slope (Current and previous waypoints have same north coordinate)
+	if (curr_wp_ned(0) - prev_wp_ned(0) < FLT_EPSILON) {
+		if (fabsf(curr_wp_ned(0) - curr_pos_ned(0)) > lookahead_distance) { // No intersection point exists
+			return bearing_to_curr_wp; // Fallback
+
+		} else {
+			const float x0 = curr_wp_ned(0) - curr_pos_ned(0);
+			const float y0 = sqrtf(powf(lookahead_distance, 2.f) - powf(x0, 2.f));
+
+			if (fabsf(curr_wp_ned(1) - (curr_pos_ned(1) + y0)) < fabsf(curr_wp_ned(1) - (curr_pos_ned(1) - y0))) {
```

Comment:
That's just to tell if we should go to the East or West intersection, right?
Then you could just say that 
```const float y0 = sign(curr_wp_ned(1) - curr_pos_ned(1)) * sqrtf(powf(lookahead_distance, 2.f) - powf(x0, 2.f));```

and only use `atan2f(y0, x0);`

Reply:
Yes, this is much cleaner thanks!
Implemented in 9e00187.